api/credentials_controller.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/gofrs/uuid"
+	"github.com/tidwall/gjson"
+	"net/http"
+	"time"
+
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+// CredentialsController implements api.Controller by providing logic for broker platform credential storage/update
+type CredentialsController struct {
+	*BaseController
+}
+
+func NewCredentialsController(ctx context.Context, options *Options) *CredentialsController {
+	return &CredentialsController{
+		BaseController: NewController(ctx, options, web.BrokerPlatformCredentialsURL, types.BrokerPlatformCredentialType, func() types.Object {
+			return &types.BrokerPlatformCredential{}
+		}),
+	}
+}
+
+func (c *CredentialsController) Routes() []web.Route {
+	return []web.Route{
+		{
+			Endpoint: web.Endpoint{
+				Method: http.MethodPut,
+				Path:   c.resourceBaseURL,
+			},
+			Handler: c.setCredentials,
+		},
+	}
+}
+func (c *BaseController) setCredentials(r *web.Request) (*web.Response, error) {
+	ctx := r.Context()
+	log.C(ctx).Debugf("Creating new broker platform credentials")
+
+	platform, err := osb.ExtractPlatformFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	body := &types.BrokerPlatformCredential{}
+	if err := util.BytesToObject(r.Body, body); err != nil {
+		return nil, err
+	}
+	body.PlatformID = platform.ID
+
+	if body.NotificationID != "" {
+		criteria := []query.Criterion{
+			query.ByField(query.EqualsOperator, "id", body.NotificationID),
+			query.ByField(query.EqualsOrNilOperator, "platform_id", body.PlatformID),
+			query.ByField(query.EqualsOperator, "resource", string(types.ServiceBrokerType)),
+		}
+
+		obj, err := c.repository.Get(ctx, types.NotificationType, criteria...)
+		if err != nil {
+			if err == util.ErrNotFoundInStorage {
+				return nil, &util.HTTPError{
+					ErrorType:   "CredentialsError",
+					Description: fmt.Sprintf("Invalid notification ID %s - request rejected", body.NotificationID),
+					StatusCode:  http.StatusConflict,
+				}
+			}
+			return nil, util.HandleStorageError(err, c.objectType.String())
+		}
+
+		notification := obj.(*types.Notification)
+
+		if gjson.GetBytes(notification.Payload, "new.resource.id").String() != body.BrokerID {
+			return nil, &util.HTTPError{
+				ErrorType:   "CredentialsError",
+				Description: fmt.Sprintf("Invalid notification ID %s - request rejected", body.NotificationID),
+				StatusCode:  http.StatusConflict,
+			}
+		}
+	}
+
+	criteria := []query.Criterion{
+		query.ByField(query.EqualsOperator, "platform_id", platform.ID),
+		query.ByField(query.EqualsOperator, "broker_id", body.BrokerID),
+	}
+	objFromDB, err := c.repository.Get(ctx, c.objectType, criteria...)
+	if err != nil {
+		if err == util.ErrNotFoundInStorage {
+			return c.registerCredentials(ctx, body)
+		}
+		return nil, util.HandleStorageError(err, c.objectType.String())
+	}
+
+	if body.NotificationID == "" {
```

Comment:
It should be like this, because in the full resync scenario we have one happy path, which is if there aren't any broker credentials we must create them, and if already exists full resync doe s not update them. This flow does not include notificationID in the body.